[9.0] [MIG+IMP] pos_customer_required

legalsylvain opened the pull request:
Hi, 
- ported the V8.0 PR #29 (pos_require_customer) from @JosDeGraeve in V9 API;
- added some features;

With this module,
- you can prevent user to start selling (front / back office) if customer is not defined;
- OR you can prevent user to make payment (front + back end) if customer is not defined;

Settings are available on each pos config.
![fromage_blanc](https://cloud.githubusercontent.com/assets/3407482/12449042/b2fcbbaa-bf7a-11e5-822e-3a25df6a0269.png)

complete description on readme.rst file.

Thanks for your review.

Kind regards.

legalsylvain committed:
[PORT] copy code from V8 PR (https://github.com/OCA/pos/pull/29) Courtesy @JosDeGraeve;
Set this module as not installable;

legalsylvain committed:
[PORT] pos_require_customer to V9;
[ADD] demo data;
[REF] pos_config.require_customer is now a selection field (instead of boolean) to allow future improvements;
[REF] OCA conventions; (one file per model);
[IMP] JS file is now loaded only in a PoS context;

legalsylvain committed:
[ADD] manage require_customer feature for front-office, if require_customer = 'order';

legalsylvain committed:
[IMP] module description;

legalsylvain committed:
[FIX] new flake8 rule;

legalsylvain committed:
[ADD] in backend pos, prevent user to make payment if customer is not set;

pedrobaeza commented:
Can you please name it _pos_customer_mandatory_ for avoiding verbs in the modules name?

JosDeGraeve commented on pos_require_customer/models/pos_config.py:
```diff
@@ -0,0 +1,32 @@
+        help="Require customer for orders in this point of sale:\n"
+        "* 'no': customer is optional;"
+        "* 'payment': customer is required before paying;"
+        "* 'order': customer is required before starting the order;")
```
It might be better to have the "user friendly" name of a selection item in the help text, like 'Optional' instead of 'no' ?

JosDeGraeve commented:
I never got around to writing js test to verify the client side functionality.  That's why the module was still in PR.

Have you planned to include js tests ?

legalsylvain committed:
[FIX] thanks pedro review.

legalsylvain commented:
@JosDeGraeve : Well, I'd like to include test, but I didnt learnt how to write such tests for the time being. That's a lot of time I have to work on this topic. I don't think it's was a blocking point for your PR. (2 :+1:  was missing).
But yes it's an important point on the "OCA / web" and "OCA / pos" roadmap.

@pedrobaeza, @JosDeGraeve : Thanks for your review. I change the description, the module name, according your remarks.

legalsylvain commented:
Hi @pedrobaeza, @JosDeGraeve,
- changed the name of the module, according the name change of the 8.0 serie;
- fixed the test;
- added a migration script from V8.0 serie;

thanks for your review.

CC  @StefanRijnhart, as you reviewed the V8.0 serie...

@gurneyalex : I used openupgradelib to make migration. Do I have to do something to make runbot works, or do you have to add a lib manually runbot ?

kind regards.

pedrobaeza commented:
The module name should be _pos_customer_required_

gurneyalex commented:
@legalsylvain I just installed the mssing dependency and forced rebuild. Let's see how it goes

legalsylvain commented:
@pedrobaeza : I renamed the module as you proposed.

@gurneyalex : Build is now ok. Thanks !

Regards.

pedrobaeza commented on pos_customer_required/migrations/9.0.2.0/post-migration.py:
```diff
@@ -0,0 +1,25 @@
+        """ % (new, openupgrade.get_legacy_name('require_customer'), old))
+
+
+def migrate(cr, installed_version):
```
Decorate with `@openupgrade.migrate()`

pedrobaeza commented on pos_customer_required/README.rst:
```diff
@@ -0,0 +1,88 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+PoS - Customer Mandatory
```
Line above

pedrobaeza commented on pos_customer_required/README.rst:
```diff
@@ -0,0 +1,88 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
```
Target link

pedrobaeza commented on pos_customer_required/README.rst:
```diff
@@ -0,0 +1,88 @@
+Usage
+=====
+
+For further information, please visit:
```
Remove this and put try me on runbot button

pedrobaeza commented on pos_customer_required/demo/pos_config.yml:
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
+##############################################################################
```
Please use header like in https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

pedrobaeza commented on pos_customer_required/migrations/9.0.2.0/post-migration.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def migrate_require_customer(cr):
+    for old, new in BOOLEAN_TO_SELECTION.iteritems():
```
You can use openupgradelib `map_values` method: https://github.com/OCA/openupgradelib/blob/4bb94fec0e3571cbc530e09b9c5b7903ece35f00/openupgradelib/openupgrade.py#L801

pedrobaeza commented on pos_customer_required/migrations/9.0.2.0/pre-migration.py:
```diff
@@ -0,0 +1,20 @@
+
+@openupgrade.migrate()
+def migrate(cr, version):
+    if not version:
```
This is not needed with the decorator

pedrobaeza commented on .travis.yml:
```diff
@@ -33,7 +33,7 @@ install:
   - travis_install_nightly
   - printf '[options]\n\nrunning_env = dev' > ${HOME}/.openerp_serverrc
-  - pip install unidecode pyserial pycountry
+  - pip install unidecode pyserial pycountry openupgradelib
```
Migration scripts are not called in Travis, so no need to add openupgradelib.

legalsylvain replied:
Travis failed because I remove openupgradelib ref. reverted.

pedrobaeza replied:
This is finally not needed, but it doesn't hurt also.

pedrobaeza commented on pos_customer_required/models/pos_config.py:
```diff
@@ -0,0 +1,26 @@
+        selection=_REQUIRE_CUSTOMER_KEYS,
+        string='Require Customer',
+        help="Require customer for orders in this point of sale:\n"
+        "* 'Optional' (customer is optional);"
```
You have missed `\n` at the end of these lines.

pedrobaeza commented on pos_customer_required/pos_customer_required.pot:
```diff
@@ -0,0 +1,62 @@
+# Translation of Odoo Server.
```
Remove the pot template

pedrobaeza commented:
In the icon, instead of "Agrolait", I would just put "Customer".

legalsylvain commented:
@pedrobaeza, @JosDeGraeve : I totally forgot this PR and your comments. 
Now it is good. 

thanks.

pedrobaeza commented:
Can you check Travis?

legalsylvain committed:
[FIX] install openupgradelib for travis

legalsylvain commented:
@pedrobaeza. Done. It's now green.

StefanRijnhart commented:
:+1: thanks

pedrobaeza commented on pos_customer_required/README.rst:
```diff
@@ -0,0 +1,92 @@
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/pos/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/{project_repo}/issues/new?body=module:%20pos_customer_required%0Aversion:%209.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Put here repo, or better switch to the new text

pedrobaeza commented on pos_customer_required/__openerp__.py:
```diff
@@ -0,0 +1,30 @@
+    ],
+    'installable': True,
+    'external_dependencies': {
+        'python': ['openupgradelib'],
```
This is why you need openupgradelib in Travis config. I don't usually put this dependency as it's only needed when updating module, not when installing. I think it's better to remove it.

legalsylvain replied:
Do you propose to try except the import ?

pedrobaeza replied:
No, you just need to remove this dependency. If someone needs to upgrade, then he/she will face the error, but only on that case.

pedrobaeza commented on pos_customer_required/__openerp__.py:
```diff
@@ -0,0 +1,30 @@
+
+{
+    'name': 'Point of Sale Require Customer',
+    'version': '9.0.2.0.0',
```
Version should be 9.0.1.0.0 (it's different from 8.0.1.0.0, so no problem with that).

pedrobaeza commented on pos_customer_required/__openerp__.py:
```diff
@@ -0,0 +1,30 @@
+    ],
+    'installable': True,
+    'external_dependencies': {
+        'python': ['openupgradelib'],
```
And more as the migration scripts will only be called in a migration from 8.0 to 9.0 via OpenUpgrade.

pedrobaeza commented on pos_customer_required/migrations/9.0.2.0/post-migration.py:
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
The folder name is incorrect

legalsylvain committed:
[FIX] remove useless openupgradelib dependencies

pedrobaeza commented:
There are still comments unattended

legalsylvain committed:
[FIX] various remarks. Courtesy pedrobaeza

pedrobaeza commented:
Please don't name GH aliases on commits, or we will be pinged each time the branch is forked, pushed and so on.

legalsylvain commented:
Ow. OK. I didn't know. without @ it's OK ?

pedrobaeza commented:
Yeah, that can serve

legalsylvain commented:
Commit ammended. regards.
